rfd/0019-cluster-routing.md

```diff
@@ -0,0 +1,73 @@
+
+This would allow users to use `tsh ssh -J clusterName serverName` to work on multiple leaf clusters concurrently.
+
+### Problems
+
+It is not immediately clear what data the jumphost address should be matched against. Clusters, and especially trusted clusters, need not be configured with options like `ssh_public_addr` or `public_addr`. Even if these options were defined there is no guarantee to their accuracy. There might be other, equally functional public addresses that can be entered by some users. It seems possible to even have multiple trusted clusters that are configured with identical `*public_addr` values (whether by accident or as part of some unconventional setup).
+
+If a jumphost input resolution fails, the client will fall back to routing through the root cluster, in spite of the user's expectation to the contrary. This may cause poor user experience as there is no easy way for an ordinary user to find out the exact leaf proxy address in the form known to the root cluster.
+
+## Alternative: `-J` with `--cluster`
+
+Another approach is based on the idea to split the problem described in the RFD into two parts. The goals are:
+
+1. to bypass the root proxy when connecting to leaf clusters; and
+2. to conveniently (yet reliably) fetch the appropriate cert for the targeted leaf cluster, without requesting repeated reissues from the root auth.
+
+Ad 1: The `-J` flag covers this aspect. Note that the jumphost is not required to be a Teleport proxy. In general it could be any other intermediary able to provide the expected routing functionality.
+
+Ad 2: See [PR #5938](https://github.com/gravitational/teleport/pull/5938/) (*Cache per-cluster SSH certificates under ~/.tsh*). SSH certs issued for the user are to be stored under `~/.tsh` and reused on an as-needed basis. The `--cluster` flag is also made to trigger SSH cert reissue when a new cluster is selected -- the cert is then loaded just for the scope of the command, i.e. without changing the cluster selected in the profile.
+
+### Problems
+
+Unfortunately, the combination of `-J` and `--cluster` does not allow simple one-liners like
+```bash
+tsh ssh --cluster=$leafCluster -J $leafProxy ...
+```
+in full generality, i.e. for an arbitrary leaf cluster right away. The problem is that the implicit cert reissue requests sent by `--cluster` are also impacted by the `-J` override. The requests are thus sent to the leaf proxy that is unable to handle such requests properly.
+
+Wouldn't it be possible to ignore `-J` for the reissue requests? Unfortunately, [some Teleport code](https://github.com/gravitational/teleport/blob/cd399e704c45f1ff8dfec6cb93262597de7ac3ba/lib/client/api.go#L1756-L1759) indicates the use of `-J` to direct `tsh` to the root proxy is indeed a deliberate feature. Can we confirm there would be a risk of breaking backward compatibility involved if `-J` were ignored? Shouldn't `--proxy` and `--user` be preferred for this purpose anyway?
```

Comment:
The fact that `-J` impacts also the way `tsh` connects to the proxy itself appears to have been implemented deliberately, otherwise the developer wouldn't have extracted proxy principals from the `-J` input. This is but an observation, not an argument that is a good design.

The point therefore is that we should take care not to break any backward compatibility even within the limited scope of cert reissual.

> If you wanted to only connect to the leaf, wouldn't you run `tsh ssh --proxy leaf.example.com node`?

A related problem (most probably a bug) is that `--proxy` isn't taken into account if there is already a proxy address saved in the profile.